hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go

```diff
@@ -173,308 +175,35 @@ func (builder *convertToARMBuilder) fixedValuePropertyHandler(propertyName astmo
 
 	fromProp, ok := fromType.Property(toProp.PropertyName())
-
-	if !ok || !toProp.PropertyType().Equals(fromProp.PropertyType()) {
+	if !ok {
 		return nil
 	}
 
-	if typeRequiresCopying(fromProp.PropertyType()) {
-		// We can't get away with just assigning this field, since
-		// it's a reference type. Use the conversion code to copy the
-		// elements.
-		source := &dst.SelectorExpr{
-			X:   dst.NewIdent(builder.receiverIdent),
-			Sel: dst.NewIdent(string(toProp.PropertyName())),
-		}
-
-		destination := &dst.SelectorExpr{
-			X:   dst.NewIdent(builder.resultIdent),
-			Sel: dst.NewIdent(string(toProp.PropertyName())),
-		}
-
-		return builder.toARMComplexPropertyConversion(
-			complexPropertyConversionParameters{
-				source:            source,
-				destination:       destination,
-				destinationType:   toProp.PropertyType(),
-				nameHint:          string(toProp.PropertyName()),
-				conversionContext: nil,
-				assignmentHandler: assignmentHandlerAssign,
-				sameTypes:         true,
-			},
-		)
-	}
-
-	result := astbuilder.SimpleAssignment(
-		&dst.SelectorExpr{
-			X:   dst.NewIdent(builder.resultIdent),
-			Sel: dst.NewIdent(string(toProp.PropertyName())),
-		},
-		token.ASSIGN,
-		&dst.SelectorExpr{
-			X:   dst.NewIdent(builder.receiverIdent),
-			Sel: dst.NewIdent(string(fromProp.PropertyName())),
-		})
-
-	return []dst.Stmt{result}
-}
-
-func (builder *convertToARMBuilder) propertiesWithSameNameButDifferentTypeHandler(
-	toProp *astmodel.PropertyDefinition,
-	fromType *astmodel.ObjectType) []dst.Stmt {
-
-	fromProp, ok := fromType.Property(toProp.PropertyName())
-	if !ok || toProp.PropertyType().Equals(fromProp.PropertyType()) {
-		return nil
+	source := &dst.SelectorExpr{
+		X:   dst.NewIdent(builder.receiverIdent),
+		Sel: dst.NewIdent(string(toProp.PropertyName())),
 	}
 
 	destination := &dst.SelectorExpr{
 		X:   dst.NewIdent(builder.resultIdent),
 		Sel: dst.NewIdent(string(toProp.PropertyName())),
 	}
 
-	source := &dst.SelectorExpr{
-		X:   dst.NewIdent(builder.receiverIdent),
-		Sel: dst.NewIdent(string(fromProp.PropertyName())),
-	}
-
-	return builder.toARMComplexPropertyConversion(
-		complexPropertyConversionParameters{
-			source:            source,
-			destination:       destination,
-			destinationType:   toProp.PropertyType(),
-			nameHint:          string(toProp.PropertyName()),
-			conversionContext: nil,
-			assignmentHandler: assignmentHandlerAssign,
-		})
-}
-
-//////////////////////////////////////////////////////////////////////////////////
-// Complex property conversion (for when properties aren't simple primitive types)
-//////////////////////////////////////////////////////////////////////////////////
-
-func (builder *convertToARMBuilder) toARMComplexPropertyConversion(
-	params complexPropertyConversionParameters) []dst.Stmt {
-
-	switch concrete := params.destinationType.(type) {
-	case *astmodel.OptionalType:
-		return builder.convertComplexOptionalProperty(params)
```

Comment:
This seems to have a built in assumption that the source and destination properties are the same type, which possibly makes this a simpler application than the conversions I've written for storage types.

Reply:
Do you specifically mean the optional property handling here? Yes, in the convert to/from ARM case optionality is preserved between the two types always (no loss of information possible).

Definitely agree that these are some hurdles we would have to cross before merging the infrastructure for these two types of handlers (storage vs ARM). Part of merging them would also probably be more fine-grained control so that each usage (storage vs arm) could easily override the bits that made sense for it.

For now though I just maintained this behavior as that's what I had before and it wasn't clear that there was a generic solution that would work for both - which is to say I suspect the storage cases where only one side is optional we'd need handlers for those specific cases (in addition to the default "both" case)?